benchmarks/apps/hpcg/hpcg.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2021 University College London (UCL) Research Software Development
+# Group.  See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import os.path as path
+import sys
+import reframe as rfm
+import reframe.utility.sanity as sn
+import glob
+
+# Add top-level directory to `sys.path` so we can easily import extra modules
+# from any directory.
+sys.path.append(path.join(path.dirname(__file__), '..', '..'))
+# `SpackTest` is a class for benchmarks which will use Spack as build system.
+# The only requirement is to inherit this class and set the `spack_spec`
+# attribute.
+from modules.utils import SpackTest
+
+# Class to define the benchmark.  See
+# https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#the-reframe-module
+# for more information about the API of ReFrame tests.
+#@rfm.simple_test
+class HPCGBenchmark(SpackTest):
+    # Systems and programming environments where to run this benchmark.  We
+    # typically run them on all systems ('*'), unless there are particular
+    # constraints.
+    valid_systems = ['*']
+    valid_prog_environs = ['default']
+    num_cpus_per_task = 1
+    # The program for running the benchmarks.
+    executable = 'xhpcg'
+    # Arguments to pass to the program above to run the benchmarks.
+    executable_opts = []
+    # Time limit of the job, automatically set in the job script.
+    time_limit = '3m'
+    # These extra resources are needed for when using the SGE scheduler.
+    extra_resources = {
+        'mpi': {'num_slots': num_tasks * num_cpus_per_task}
+    }
+
+    reference = {
+        # flops? Gflops/s ?
+        'archer2': {
+            'flops': (1000.0, -0.2, None, 'Gflops/seconds'),
+        },
+        '*': {
+            'flops': (1, None, None, 'Gflops/seconds'),
+        }
+    }
+
+    @run_after('setup')
+    def setup_variables(self):
+        # Strictly HPCG is only intended to run for 1 OMP thread, except in original version
+        self.variables['OMP_NUM_THREADS'] = f'{self.num_cpus_per_task}'
+
+    @run_before('compile')
+    def setup_build_system(self):
+        self.build_system.specs = [self.spack_spec]
+
+    @run_before('sanity')
+    def set_sanity_patterns(self):
+        # Should probably check that it's a valid run
+        self.sanity_patterns = sn.assert_found(r'', self.stdout)
```

Comment:
Maybe this can check something more useful than nothing?  For example, is there anything written to stdout/stderr that could be useful to match?

Reply:
There is something in the self.perf_patterns which looks for the word 'Valid ....' in the output file. This could be used again (check only one possible_outfiles too)